dts: bindings: clean up redundant required false attributes

cfriedt opened the pull request:
DTS property attributes are (by default) not required.

Explicitly specifying `required: false` is redundant. Perhaps a warning to that effect would be useful.

Note: this is a non-breaking tree-wide change, and should follow that process.

Commit:
dts: bindings: clean up redundant required false attributes

DTS property attributes are (by default) not required.

Explicitly specifying `required: false` is redundant.
Perhaps a warning to that effect would be useful.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

gmarull reviewed (dismissed):
A CI check would be nice. Otherwise, this will proliferate again.

pdgendt commented:
A warning would be nice indeed, there are some open PRs with `required: false` entries:
#46134 #45370 #42570

cfriedt commented:
I'll add a CI check in a subsequent commit

pdgendt commented on dts/bindings/timer/st,stm32-timers.yaml:
```diff
@@ -15,7 +15,6 @@ properties:
       required: true
 
     interrupts:
-      required: false
```
Need to check for empty entries, but more importantly, it **can** be that you override a base `required: true` property which makes it not redundant.

cfriedt replied:
Hmm... it might potentially be better to add an option to warn for redundant `required: falsee` into `dtc`, unless we have access to the full AST from python.

gmarull replied:
Hmm.. I wonder if it is a good idea to actually override a default required true... It may have some usecases though.

mbolivar-nordic replied:
Wait, what do you mean? You can't "un-require" a property.

galak reviewed (commented):
leave the ones in base.yaml for documentation purposes.

Commit:
scripts: compliance: add devicetree binding compliance check

Add `DevicetreeBindingsCheck` and `required_false_check()`, which
checks to make sure that no new Devicetree bindings add the
redundant `required: false` property.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
* moved CI check to `scripts/ci/check_compliance.py`
* rebased on main

cfriedt commented:
Since I created this PR and now, there have already been 13 DT bindings with `required: false` added 🤣

Commit:
doc: build: dts: bindings: discourage use of redundant: false

Aside from base properties (maybe for documentation purposes),
`required: false` should not be encouraged.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
* updated docs to discourage use of `required: false`
* removed *more* newly-added `required: false` properties in dt bindings

cfriedt commented:
Re-request review is broken in GH right now. Ping @pdgendt, @galak, @gmarull

cfriedt commented:
ping @carlescufi @stephanosio

mbolivar-nordic commented on scripts/ci/check_compliance.py:
```diff
@@ -220,6 +220,50 @@ def run(self):
                 self.failure(output)
 
 
+class DevicetreeBindingsCheck(ComplianceTest):
```
I'm not going to NAK over this because this is already a very altruistic pull request, and this kinda works, but it is duplicating features that we already have elsewhere in a slightly worse way.

See e.g.

```py
    bindings = edtlib.bindings_from_paths(binding_files, ignore_errors=True)
```

in gen_devicetree_rest.py, which relies on the edtlib abstraction boundary to find the bindings, and supports YAML files with both .yml and .yaml extensions.

If you used this, you'd get a bunch of first-class `Binding` objects, and you could use `binding.raw` values to look and see if anything in `raw['properties']` has an explicit `required: true` instead of just checking if the string `"required: false"` appears in the text of the file. The latter string-based check is subject to false positives if the string appears in a YAML comment, for example.

Something to improve in the future, though, if we just want to get this in now.
